File: Depth-First Search/Binary tree/Find paths/Hard/LET_no124_Binary Tree Maximum Path Sum.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: The root of binary tree.
    @return: An integer
    """
    def maxPathSum(self, root):

        maxSum, _ = self.dfs(root)
        return maxSum

    def dfs(self, root):
        if not root:
            return -sys.maxsize, 0

        left_max, left_sum = self.dfs(root.left)
        right_max, right_sum = self.dfs(root.right)
        maxSum = max(left_max, right_max, left_sum + root.val + right_sum)
        singleSum = max(left_sum + root.val, right_sum + root.val, 0)

        return maxSum, singleSum

File: Depth-First Search/Binary tree/Find paths/Hard/LET_no124_Binary Tree Maximum Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int res = INT_MIN;
public:
    int maxPathSum(TreeNode* root) {
        helper(root);
        return res;
    }

    int helper(TreeNode* cur) {
        if (!cur)
            return 0;

        int left_sum = helper(cur->left), right_sum = helper(cur->right);
        res = max(res, left_sum + right_sum + cur->val);
        return max(0, max(left_sum + cur->val, right_sum + cur->val));
    }
};